### PR TITLE
PP-339: Node Ramp Down Feature

### PR DESCRIPTION
**Pull request by bayucan**
This is the implementation of the node rampdown feature, where no longer needed sister nodes can be released early from a job. 

#### Issue-ID

* **[PP-339](https://pbspro.atlassian.net/browse/PP-339)**

#### Problem description

This is a new feature of PBS where sister nodes can be released early using the command 'pbs_release_nodes' and also using the qsub -W release_nodes_on_stagout=true option.
#### Cause / Analysis
When sister nodes are released early, ensure that before the server assigns exclusive nodes to other jobs, ensure that all processes belonging to the job on the nodes are killed, and a summary of resources used is sent to the mother superior to be accumulated at the end of the job.
#### Solution description
The solution implements the EDD in:
https://pbspro.atlassian.net/wiki/display/PD/PP-339+and+PP-647%3A+node+ramp+down+feature%3A+release+vnodes+early+from+running+jobs
and discussion forum is in:
http://community.pbspro.org/t/pp-339-and-pp-647-release-vnodes-early-from-running-jobs
The implementation involves the addition of pbs_release_nodes command, the protocol PBS_BATCH_RelnodesJob, inter mom request IM_DELETE_JOB2, new accounting records 'u' (update), 'c' (continue), and additional internal attributes exec_host*, exec_vnode*, Resource_List_* that are not shown unless a server attribute 'show_hidden_attribs' is set to True.

#### Checklist:

- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [X] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/PD/PP-339+and+PP-647%3A+node+ramp+down+feature%3A+release+vnodes+early+from+running+jobs)** for details.
- [X] I have added documentation in the **(https://pbspro.atlassian.net/wiki/display/PD/PP-339+and+PP-647%3A+node+ramp+down+feature%3A+release+vnodes+early+from+running+jobs)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
PTL testsuite is coming...
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [ ] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

**Comment by bhroam** [comment]
Travis failed because your commit isn't signed.  Make sure to sign your commits in the future.

File: src/server/job_func.c [hunk 1]
```diff
@@ -1192,6 +1198,182 @@ decode_project(struct attribute *patr, char *name, char *rescn, char *val)
...
+	}
+
+
+	return (0);
+
+update_resources_list_error:
+	job_attr_def[backup_res_list_index].at_free(
+			&pjob->ji_wattr[backup_res_list_index]);
+	pjob->ji_wattr[backup_res_list_index].at_flags &= \
+						~ATR_VFLAG_SET;
```
Comment (bhroam) [inline src/server/job_func.c]: You're not saving much by breaking this into 2 lines.  Please combine them

File: src/server/svr_jobfunc.c [hunk 2]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+								&ns_malloced,
+								":") == NULL) {
+								log_err(-1, id, "pbs_strcat failed");	
+								goto recreate_exec_vnode_exit;
+							}
+							if (pbs_strcat(
+								&new_select,
+								&ns_malloced,
+								extra_res) == NULL) {
+								log_err(-1, id, "pbs_strcat failed");
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Please combine the multiple line pbs_strcat() calls into fewer lines.  They are hard to read at 4 lines each.

File: src/server/svr_jobfunc.c [hunk 3]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+							}
+						}
+						if (pbs_strcat( &new_select,
+							&ns_malloced,
+								"1") == NULL) {
+							log_err(-1, id,
+							"pbs_strcat failed");
+							goto recreate_exec_vnode_exit;
+						}
+						if (pbs_strcat( &new_select,
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Remove the space after the '('

File: src/server/svr_jobfunc.c [hunk 4]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+	if (new_select[entry] == '+')
+		new_select[entry] = '\0';
+
+	entry = strlen(deallocated_execvnode)-1;
+	if (deallocated_execvnode[entry] == '+')
+		deallocated_execvnode[entry] = '\0';
+
+	if (new_exec_vnode[0] != '\0') {
+
+		if (strcmp(\
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: No need for the \\

File: src/server/svr_jobfunc.c [hunk 5]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+		 pjob->ji_wattr[(int) JOB_ATR_exec_vnode].at_val.at_str,
+						 new_exec_vnode) == 0) {
+			/* no change */
+			LOG_EVENT_BUF_ARG1(err_msg,err_msg_sz-1,
+				"these nodes are not part of the job: %s",
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Maybe rephrase this as "nodes requested to be released do not belong to job %s"

File: src/server/svr_jobfunc.c [hunk 6]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+				(char *)0,
+				(char *)0,
+				pjob->ji_wattr[(int) JOB_ATR_exec_vnode].at_val.at_str);
+		}
+
+		if ((pjob->ji_wattr[JOB_ATR_resource_acct].at_flags & ATR_VFLAG_SET) != 0) {
+			job_attr_def[JOB_ATR_resource_acct].at_free(&pjob->ji_wattr[JOB_ATR_resource_acct]);
+			pjob->ji_wattr[JOB_ATR_resource_acct].at_flags &= ~ATR_VFLAG_SET;
+		}
+		job_attr_def[JOB_ATR_resource_acct].at_set( &pjob->ji_wattr[JOB_ATR_resource_acct], &pjob->ji_wattr[JOB_ATR_resource], INCR);
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Remove the space after the '('

File: src/server/svr_jobfunc.c [hunk 7]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+							      prdefsl);
+			if (presc == NULL) {
+				presc = add_resource_entry(
+			  	 &pjob->ji_wattr[(int)JOB_ATR_resource],
+							prdefsl);
+			}
+			if (presc != NULL) {
+				(void)prdefsl->rs_decode(
+					&presc->rs_value, NULL,
+						"select", new_select);
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: The indent is off.  They cascade.  Please make all lines uniform

File: src/server/svr_jobfunc.c [hunk 8]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+			/* already saved in *_orig */
+			if ((pjob->ji_wattr[JOB_ATR_SchedSelect_orig].\
+				      at_flags & ATR_VFLAG_SET) == 0) {
+				(void)decode_str(
+                         		&pjob->ji_wattr[\
+					(int)JOB_ATR_SchedSelect_orig],
+					NULL,
+                                        NULL, 
+					pjob->ji_wattr[\
+				    JOB_ATR_SchedSelect].at_val.at_str);
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Can you combine a few of these lines?  Plus you don't need the \\s

File: src/server/svr_jobfunc.c [hunk 9]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+		char	*tmpbuf2 = NULL;
+		char	*pc = NULL;
+		char	*pc1 = NULL;
+
+		tmpbuf = strdup(vnodelist);
+		tmpbuf2 = strdup(vnodelist); 	/* will contain nodes */
+					    	/* that are in */
+						/* 'vnodelist' but */
+						/* not in */
+						/* deallocated_execvnode */
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Please combine some this comment to fewer lines

File: src/server/svr_jobfunc.c [hunk 10]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+				while (prs) {
+					if ( (prdefvntype != NULL) &&
+						(prs->rs_defin == prdefvntype) &&
+						(prs->rs_value.at_flags & ATR_VFLAG_SET) != 0 ) {
+						struct array_strings *as;
+						int	l;
+						as = prs->rs_value.at_val.at_arst;
+						for (l = 0; l < as->as_usedptr; l++) {
+							if ((strcmp(as->as_string[l], CRAY_LOGIN) == 0) ||
+								(strcmp(as->as_string[l], CRAY_COMPUTE) == 0)) {
```
Comment (lisa-altair) [inline src/server/svr_jobfunc.c]: Thanks @bhroam for bringing this to my attention.  I agree with the reasoning behind your suggestion.  But I fear it may be a bit complex.  I have asked a few questions to @bayucan in the forum related to the external design for this feature.  If it is expected that linux systems managed by the same PBS that is managing a Cray X* series system, then @bhroam code suggestion of strstr for "cray" could prevent those linux systems from being ramped down (see 3a of my forum comment).  Of course this "cray" keyword could be called out in the external design, so customers that want to use node ramp down will know what NOT to set for vntype.  This may be a good idea anyway since vntype may be set by admins.

File: src/server/svr_jobfunc.c [hunk 11]
```diff
@@ -7108,14 +7085,81 @@ recreate_exec_vnode(job *pjob, char *vnodelist, char *err_msg,
 	if (deallocated_execvnode[entry] == '+')
 		deallocated_execvnode[entry] = '\0';
 
+	if (deallocated_execvnode[0] != '\0') {
```
Comment (bayucan) [inline src/server/svr_jobfunc.c]: Moved up these series of calls so that if something fails in them, non of the  job attributes (exec_vnode, exec_host, etc...) would have been set yet.

File: src/server/node_manager.c [hunk 12]
```diff
@@ -7870,6 +7890,11 @@ update_job_node_rassn(job *pjob, attribute *pexech, enum batch_op op)
 				prdef = find_resc_def(svr_resc_def, pkvp[j].kv_keyw, svr_resc_size);
 				if (prdef == NULL)
 					return;
+	
+				if ((prdef->rs_flags & asgn) == 0) {
```
Comment (bayucan) [inline src/server/node_manager.c]: This is the general fix to the crash seen at NASA due to presence of "aoe" in exec_vnode. This resource is not flagged to be set under resources_assigned. So the flag is check here to make sure it has a a flag of ANASSN or FNASSN and if not, just continue.

File: src/server/node_manager.c [hunk 13]
```diff
@@ -2689,6 +2690,326 @@ recv_wk_job_idle(int stream)
...
+	jobid = pjob->ji_qs.ji_jobid;
+	if (*jobid == '\0')
+		return;
+
+	for (i = 0; i < svr_totnodes; i++) {
+		pnode = pbsndlist[i];
+
+		if ((pnode != NULL) && !(pnode->nd_state & INUSE_DELETED)
+			&& (pnode->nd_moms != NULL) &&
+				(pnode->nd_moms[0] == pmom)) {
```
Comment (bhroam) [inline src/server/node_manager.c]: I'm not sure I understand.  My question was about your pnode->nd_moms[0].  You're only looking at the 0th entry in the array.  There are node's that have multiple moms.  I believe this is only used on the cray, so you're fine there.  My point is that if that you start supporting cray in the future or other systems start using the multi-mom feature, this could easily be missed.

Comment (bayucan) [inline src/server/node_manager.c]: I've fixed this in the latest code line. I'm now going through all the parent moms and looking for a match.

File: src/server/node_manager.c [hunk 14]
```diff
@@ -8007,3 +8521,82 @@ req_momrestart(struct batch_request *preq)
...
+	rc = send_job_exec_update_to_mom(pjob, err_msg, err_msg_sz,
+							reply_req);
+
+	if ( rc == 0) {
+
+		if ((pjob->ji_qs.ji_state == JOB_STATE_RUNNING) &&
+		    (pjob->ji_qs.ji_substate == JOB_SUBSTATE_RUNNING)) {
+			/* job must be in a running state */
+			/* so as to not conflict with the */
+			/* account end record */
```
Comment (bhroam) [inline src/server/node_manager.c]: If I was a bit confused, you might want to reword the comment.

File: src/server/svr_jobfunc.c [hunk 15]
```diff
@@ -5723,3 +5724,1613 @@ svr_avljob_oper(job *pjob, int delkey)
...
+ *	then this function returns:
+ *		<resC>=<valC>:<resD>=<valD>
+ *
+ * @param[in]	res_list - the resources list
+ * @param[in]	chunk - the chunk to check for new resources.
+ *
+ * @return char *
+ * @retval != NULL	the resources that are used in 'chunk',
+ *			but not in 'res_list'.
+ * @retval == NULL	if error encountered.
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: You might want to add that to the doxygen comment

**Comment by bhroam** [comment]
@bayucan, Travis failed again due to checksign.  It isn't mandatory to have travis pass until we're ready to merge, but it is nice to know that it will.  Checksign is run before the tests are, so there may be hidden test failures.

**Comment by mike0042** [comment]
If you want to sign your commits for the two changes you've made, do a "git rebase -i HEAD~2" and add the following line after each pick:
exec git commit --amend --no-edit -S
Then do a force push to your dev branch.

File: src/server/svr_jobfunc.c [hunk 16]
```diff
@@ -6436,26 +6436,25 @@ resources_seen(char *exec_vnode)
...
 return_missing_resources(char *chunk, char *res_list)
 {
-	static		char *id = "return_missing_resources";
 	int             snc;
  	int             snelma;
 	static int	snelmt = 0;       /* must be static per parse_chunk_r() */
 	static key_value_pair *skv = NULL; /* must be static per parse_chunk_r() */
-        int             rc = 0;
-	static		char	ret_buf[LOG_BUF_SIZE-1] = {0};
+        int		rc = 0;
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Is there an extra space in front of int?

File: src/server/svr_jobfunc.c [hunk 17]
```diff
@@ -6468,37 +6467,27 @@ return_missing_resources(char *chunk, char *res_list)
...
 				buf_len = strlen(ret_buf);
-				strncat(ret_buf,
-				   "=",
-				   buf_sz-buf_len-1);
+				strncat(ret_buf, "=", buf_sz-buf_len-1);
 				buf_len += 1;
-				strncat(ret_buf,
-				   skv[l].kv_val,
-				   buf_sz-buf_len-1);
+				strncat(ret_buf, skv[l].kv_val, buf_sz-buf_len-1);
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Only the first character of ret_buf gets initialized to '\0'.  You need to handle the case when strncat() doesn't terminate the string.

File: src/server/svr_jobfunc.c [hunk 18]
```diff
@@ -6508,15 +6497,46 @@ return_missing_resources(char *chunk, char *res_list)
...
+ *	This return 1 if the given 'host' and 'part' matches the
+ *	parent mom of node 'pnode'.
+ *
+ * @param[in]	pnode - the node to match host against
+ * @param[in]	host - hostname to match
+ * @param[in]	port - port to match
+ *
+ * @return int
+ * @retval 1	- if true
+ * @retval 0    - if  false
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Is there an extra space before the hyphen?

File: src/server/svr_jobfunc.c [hunk 19]
```diff
@@ -6588,49 +6607,49 @@ recreate_exec_vnode(job *pjob, char *vnodelist, char *err_msg,
 	resource_def	*prdefvntype;
 	resource_def	*prdefarch;
 
-	if ( pjob == NULL ) {
-		LOG_ERR_BUF(err_msg, err_msg_sz-1, "bad job parameter",
-								-1, id)
+	if (pjob == NULL) {
+		LOG_ERR_BUF(err_msg, err_msg_sz, "bad job parameter",
+								-1, __func__)
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Can you combine the above two lines?  You're not saving much by breaking them.

File: src/server/svr_jobfunc.c [hunk 20]
```diff
@@ -6805,19 +6820,17 @@ recreate_exec_vnode(job *pjob, char *vnodelist, char *err_msg,
 				}
 			}
 
-			if (((strcmp(pnode->nd_moms[0]->mi_host,
-						ms_fullhost) == 0) &&
-			     (pnode->nd_moms[0]->mi_port == ms_port)) ||
-			     ((vnodelist != NULL) &&	
+			if (is_parent_host_of_node(pnode, ms_fullhost, ms_port) ||
+			     ((vnodelist != NULL) &&
 			      !in_string_list(noden, '+', vnodelist))) {
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Can you combine the above two lines?  The line won't be much longer than the one above it.

File: src/server/svr_jobfunc.c [hunk 21]
```diff
@@ -7028,41 +7018,28 @@ recreate_exec_vnode(job *pjob, char *vnodelist, char *err_msg,
...
 							"pbs_strcat failed");
 							goto recreate_exec_vnode_exit;
 						}
-						if (pbs_strcat( &new_select,
-							&ns_malloced,
-								buf_sum) == NULL) {
-							log_err(-1, id,
+						if (pbs_strcat(&new_select, &ns_malloced, buf_sum) == NULL) {
+							log_err(-1, __func__,
 							"pbs_strcat failed");
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Combine the above two log_err() calls into one line each.  They won't be longer than the if() statement.

File: src/server/svr_jobfunc.c [hunk 22]
```diff
@@ -7108,14 +7085,81 @@ recreate_exec_vnode(job *pjob, char *vnodelist, char *err_msg,
...
+		char	*pc1 = NULL;
+
+		tmpbuf = strdup(vnodelist);
+		tmpbuf2 = strdup(vnodelist); 	/* will contain nodes that are in */
+						/* 'vnodelist' but not in deallocated_execvnode */
+		if ((tmpbuf != NULL) && (tmpbuf2 != NULL)) {
+
+			tmpbuf2[0] = '\0';
+
+			pc = strtok(tmpbuf, "+");
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Consider using strtok_r() instead.

File: src/server/svr_jobfunc.c [hunk 23]
```diff
@@ -7329,7 +7298,7 @@ recreate_exec_vnode(job *pjob, char *vnodelist, char *err_msg,
 	free(res_in_exec_vnode);
 	/* clear the summation buffer */
 	(void)resc_sum_values_action(RESC_SUM_GET_CLEAR, NULL, NULL, NULL, buf,
-									sizeof(buf)-1);
+									sizeof(buf));
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: If you want to split the above line into two, consider breaking it up more evenly (or just combine it into one line).

File: src/server/svr_jobfunc.c [hunk 24]
```diff
@@ -6910,68 +6921,49 @@ recreate_exec_vnode(job *pjob, char *vnodelist, char *err_msg,
...
-							"pbs_strcat failed");
+						if (pbs_strcat(&new_select, &ns_malloced, buf_sum) == NULL) {
+							log_err(-1, __func__, "pbs_strcat failed");
 							goto recreate_exec_vnode_exit;
 						}
 						if ((extra_res != NULL) && (extra_res[0] != '\0')) {
-							if (pbs_strcat(
-								&new_select,
-								&ns_malloced,
+							if (pbs_strcat(&new_select, &ns_malloced,
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Either combine the two lines of this this pbs_strcat() or break it at the '=='. I'd rather see what you're strcating on the same line.

File: src/server/node_manager.c [hunk 25]
```diff
@@ -8668,22 +8670,14 @@ free_sister_vnodes(job *pjob, char *vnodelist, char *err_msg,
 	}
 	/* increment everything found in new exec_vnode */
 	set_resc_assigned((void *)pjob, 0,  INCR);
-
+						
+	set_scheduler_flag(SCH_SCHEDULE_TERM);
 	rc = send_job_exec_update_to_mom(pjob, err_msg, err_msg_sz,
```
Comment (bayucan) [inline src/server/node_manager.c]: It turned out that I didn't need to enclose the PBS_ACCT_UPDATE and PBS_ACT_NEXT records in an if loop to apply only to running jobs..  I put the if condition here because I was worried that the records would also show up with regular jobs that didn't release any nodes. But that cannot happen since this is the part of the code inside free_sister_nodes() which is only done when pbs_release_nodes or -Wrelease_nodes_on_stageout is called/set.

File: src/server/node_manager.c [hunk 26]
```diff
@@ -8668,22 +8670,14 @@ free_sister_vnodes(job *pjob, char *vnodelist, char *err_msg,
...
 							reply_req);
 
 	if (rc == 0) {
-
-		if ((pjob->ji_qs.ji_state == JOB_STATE_RUNNING) &&
-		    (pjob->ji_qs.ji_substate == JOB_SUBSTATE_RUNNING)) {
-			/*
-			 * job must be in a running state
-			 * so as to not conflict with the
```
Comment (bayucan) [inline src/server/node_manager.c]: It turned out that I didn't need to enclose the PBS_ACCT_UPDATE and PBS_ACT_NEXT records in an if loop to apply only to running jobs.. I put the if condition here because I was worried that the records would also show up with regular jobs that didn't release any nodes. But that cannot happen since this is the part of the code inside free_sister_nodes() which is only done when pbs_release_nodes or -Wrelease_nodes_on_stageout is called/set.

File: src/server/svr_jobfunc.c [hunk 27]
```diff
@@ -5783,20 +5782,20 @@ find_ms_full_host_and_port(job *pjob, int *port)
...
 					id, err_msg); \
 }
 
 #define	LOG_EVENT_BUF_ARG3(err_msg, err_msg_sz, msg, arg1, arg2, arg3, id) \
 if ((err_msg != NULL) && (err_msg_sz > 0)) { \
-	snprintf(err_msg, err_msg_sz-1, msg, arg1, arg2, arg3); \
+	snprintf(err_msg, err_msg_sz, msg, arg1, arg2, arg3); \
 	log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_JOB, LOG_DEBUG, \
 					id, err_msg); \
 }
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Oh wait, is that the guts LOG_EVENT_BUF_ARG3 macro?  Then yes, you absolutely need them.  I thought it was just random code with \\'s at the end.

File: src/cmds/pbs_release_nodes.c [hunk 28]
```diff
@@ -171,9 +178,9 @@ char **envp;
 		if (stat && (pbs_errno == PBSE_UNKJOBID)) {
 			if (locate_job(job_id_out, server_out, rmt_server)) {
 				/*
-                                 *job located at a different server
+				 * job located at a different server
 				 * retry connect on the new server
-                                 */
+				*/
```
Comment (bhroam) [inline src/cmds/pbs_release_nodes.c]: Align the last */ so the *'s line up

File: src/server/svr_jobfunc.c [hunk 29]
```diff
@@ -6045,31 +6045,34 @@ enum resc_sum_action {
...
- *			  			 ":<res>=<value>:<res1>=(value1>:<res2>=<value2>..."
- * @retval	NULL		If an error has occurred, filling in the 'err_msg' with the error message.
- * @Par	MT-safe: No
+ * @retval	<string> If 'action' is RESC_SUM_ADD, then this returns the 'keyw' to
+ *			signal success adding the <resc_def, keyw, value>.
+ *			If 'action' is RESC_SUM_GET_CLEAR, then this returns the
+ *			<res>=<value> entries in the internal resc_sum_values
+ *			array, as well as clear/initialize entries in the resc_sum_values
+ *			array. The returned string is of the form:
+ *				":<res>=<value>:<res1>=(value1>:<res2>=<value2>..."
```
Comment (bhroam) [inline src/server/svr_jobfunc.c]: Consider aligning the line extensions with the 'If'

File: test/tests/functional/pbs_node_rampdown.py [hunk 30]
```diff
@@ -0,0 +1,13048 @@
...
+from tests.functional import *
+
+
+class TestPbsNodeRampDown(TestFunctional):
+
+    """
+    This tests the Node Rampdown Feature,
+    where while a job is running, nodes/resources
+    assigned by non-mother superior can be released.
+    Date: Aug 28 2014
```
Comment (bhroam) [inline test/tests/functional/pbs_node_rampdown.py]: There's no need to date the test.  I'd remove it.

File: src/resmom/mom_comm.c [hunk 31]
```diff
@@ -4665,25 +4663,32 @@ im_request(int stream, int version)
...
 			BAIL("resources_used.cput")
 			secs = pjob->ji_resources[resc_idx].nr_cput;
-			hours = secs/3600;
-			secs -= hours*3600;
-			mins = secs/60;
-			secs -= mins*60;
+			hours = secs / 3600;
+			secs -= hours * 3600;
+			mins = secs / 60;
+			secs -= mins * 60;
```
Comment (bayucan) [inline src/resmom/mom_comm.c]: Done.

**Comment by bhroam** [comment]
OK, because github is horribly broken and didn't keep my comments on the same line as I put them on, I'm going to have to rereview your entire test.  Let's fix it up first before that though.  There are some simple things you can do to probably cut the test size in half.  
* You recreate a bunch of strings over and over again to search for them in the job and accounting logs (like the *_esc strings).  Some of them are new (like newsel_esc), but others are created directly from the class variables.  Just create new class variables for the strings that don't change.  
* The largest bang for your buck will be the accounting log matches.  You're spending half your tests doing the same accounting log match for 4 different accounting log records.  Put the code in a function that takes all 7-8 different arguments.  This should cut your test size down by half.  As a bonus if the accounting log records ever change, you have one place to make the change instead of once per test.  
* One small thing is that when you check the logs for seeing the job start (checking for pid on MS and JOIN JOB on the sisters).  Put that into a function.  It won't save much, but in a 13,500 line test with so much duplication, every little bit helps.  
* Move the license check into a function of its own and take the number of licenses used as an argument
* Look through the test.  You have many occurrences of for loops with a single entity list.  No need for the loop.  
* There are many tests where the same expect is done in 2-3 loops.  Combine those loops.  
* You do a HUGE amount of string formatting.  The vast majority of it is adding the vnode suffixes (e.g., [0], etc) to the hosts.  Do that once and make a set of self.vnodeN class variables.  Not only will that cut down on the string manipulation, it'll make it easier to keep track of which vnode is which.  
* Speaking of string formatting, there are times where you do "%s" % (str).  There's no real need for that, just use the string directly.  
* Continuing on that, there are times you use string formatting to add static strings (i.e., "%s" % "scatter").  Don't do that, just add the static string directly.  
* Review all of your comments in your tests.  A lot of them are wrong in some way or another.  They're copy paste issues from previous tests (e.g., most of the comments above when you call pbs_release_nodes say you're using pbs_release_nodes -a).  
* You create the string for the pbs_release_nodes command(i.e., os.path.join()) each time you use it (some times multiple times in a test).  Create it once and shove it into a class variable.
* Remove the log message for the pbs_release_nodes command.  That will be printed into the log at INFOCLI2 when the command is run.
* No need to say max_attempts=1 in log_match().  That is the default.  Also, setting interval with max_attempts=1 is useless.  Setting the interval sets the amount of time to wait between checks.  You are only looking once.
* Could you revamp your doc strings?  I don't find the exec_vnode helpful.  Why don't you say "vnode's N-M are part of super-chunk 1" and so on.  Use the new vnode numbers from the class variables.  Second, you explain what your test is doing, but not why.  Many times I was confused on how the tests differed from each other.  The first few tests when you start releasing nodes look the same to me.
* Why does adding mpiprocs and ompthreads into the mix help?  Please explain this in the doc string
* Many tests do an expect() at the start and the end for the place spec.  Checking it at the start is fine, but I don't see any reason to check it again at the end.  It won't change during the test.  Speaking of the place statement, many times when you're checking for it in the accounting logs, you use a quoted string.  It's saved as a class variable.  You submitted the job using the class variable.  Always check using the class variable.
* Many times you use assertEquals(A == B).  Please use assertEquals(A, B)
* Look at PP-861.  You run into this bug in many exec_hosts in your test.  Once it is fixed, your tests will start failing.  You should do a regex match on exec_host=host/N*\d
* Many places you have log_match() with rexexp=True even though you're doing a non-regex string search.  Please remove regexp=True from all of these
* Since your test takes so long to run, one way you can help that is by increasing the max_attempts and lowering the interval in your log_match() calls.  Each time you make a long interval, there is a chance the test can wait the entire interval to find the match.  Individually it isn't that long, but they add up.
* Whenever you submit a job and expect it to be queued, do a self.server.expect(JOB, 'comment', op=SET, id=jid) first.  Just checking job state has a race condition.  The job is in the queued state before the scheduler considers it.  Your expect could pass right before the scheduler runs the job.
* In several tests you query proc info for *fib*, but don't actually do anything with the data.
* Speaking of fib, you never check for cput time, so why use fib over a plain sleep?
* Several places you check if the jid that submit() returns is None.  There is no need to do this.  The submit() call will raise an exception if it fails.

File: src/cmds/pbs_release_nodes.c (hunk 28 above)
Comment (bayucan) [inline src/cmds/pbs_release_nodes.c]: Fixed.

File: src/server/svr_jobfunc.c (hunk 29 above)
Comment (bayucan) [inline src/server/svr_jobfunc.c]: Fixed.
